File: CS/Pivot-Aggr-Examples/FileParser.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Pivot_Aggr_Examples
{
    class CodeFileParser
    {

        public static Dictionary<string, List<string>> ParseCSharpFile(string fileName)
        {
            FileInfo f = new FileInfo(fileName);
            Dictionary<string, List<String>> snippets = new Dictionary<string, List<string>>();
            string[] originCodeLines = File.ReadAllLines(fileName);
            List<string> codeLines = null;

            for (int i = 0; i < originCodeLines.Length; i++)
            {
                string sLine = originCodeLines[i];
                if (sLine != null)
                {
                    if (sLine.TrimStart().StartsWith("#region", StringComparison.CurrentCultureIgnoreCase))
                    {
                        string key = sLine.TrimStart().Substring("#region".Length).TrimStart();
                        if (snippets.ContainsKey(key))
                        {
                            throw new ArgumentException(string.Format("Duplicate names: {0}", sLine.Trim()));
                        }
                        snippets[key] = new List<string>();
                        codeLines = snippets[key];
                    }
                    else
                    {
                        if (codeLines != null)
                        {
                            if (sLine.TrimStart().StartsWith("#endregion", StringComparison.CurrentCultureIgnoreCase))
                            {
                                codeLines = null;
                            }
                            else
                            {
                                codeLines.Add(sLine.Replace(new string(' ', 12), ""));
                            }
                        }
                    }
                }
            }

            return snippets;
        }
    }
}

File: VB/Pivot-Aggr-Examples/FileParser.vb
Imports System
Imports System.Collections.Generic
Imports System.IO
Imports System.Linq
Imports System.Text
Imports System.Text.RegularExpressions
Imports System.Threading.Tasks

Namespace Pivot_Aggr_Examples
    Friend Class CodeFileParser

        Public Shared Function ParseVbFile(ByVal fileName As String) As Dictionary(Of String, List(Of String))
            Dim f As New FileInfo(fileName)
            Dim snippets As New Dictionary(Of String, List(Of String))()
            Dim originCodeLines() As String = File.ReadAllLines(fileName)
            Dim codeLines As List(Of String) = Nothing

            For i As Integer = 0 To originCodeLines.Length - 1
                Dim sLine As String = originCodeLines(i)
                If sLine IsNot Nothing Then
                    Dim sLineTrimmed As String = sLine.TrimStart(" "c, "'"c, Convert.ToChar(vbTab))
                    If sLineTrimmed.StartsWith("#Region", StringComparison.CurrentCultureIgnoreCase) Then
                        Dim key As String = sLineTrimmed.Substring("#Region ".Length).TrimStart(" "c, """"c).TrimEnd(" "c, """"c)
                        If snippets.ContainsKey(key) Then
                            Throw New ArgumentException(String.Format("Duplicate names: {0}", sLine.Trim()))
                        End If
                        snippets(key) = New List(Of String)()
                        codeLines = snippets(key)
                    Else
                        If codeLines IsNot Nothing Then
                            If sLineTrimmed.StartsWith("#End Region", StringComparison.CurrentCultureIgnoreCase) Then
                                codeLines = Nothing
                            Else
                                codeLines.Add(sLine.Replace(vbTab, " "))
                            End If
                        End If
                    End If
                End If
            Next i

            Return snippets
        End Function
    End Class
End Namespace